### PR TITLE
[Impeller] Make matrix image filter work as expected

### PR DESCRIPTION
**Pull request by ColdPaleLight**
fix https://github.com/flutter/flutter/issues/111737

#### without patch
dtV4mRtY9J

#### with patch
2qcLgw2TDv

## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [Flutter Style Guide] and the [C++, Objective-C, Java style guides].
- [ ] I listed at least one issue that this PR fixes in the description above.
- [x] I added new tests to check the change I am making or feature I am adding, or Hixie said the PR is test-exempt. See [testing the engine] for instructions on
writing and running engine tests.
- [ ] I updated/added relevant documentation (doc comments with `///`).
- [x] I signed the [CLA].
- [x] All existing and new tests are passing.

### CODE REVIEW

File: impeller/entity/contents/filters/matrix_filter_contents.cc [hunk 1]
```diff
@@ -25,7 +24,35 @@ std::optional<Snapshot> MatrixFilterContents::RenderFilter(
...
-  return inputs[0]->GetSnapshot(renderer, entity);
+  auto snapshot = inputs[0]->GetSnapshot(renderer, entity);
+  if (!snapshot.has_value()) {
+    return std::nullopt;
+  }
+
+  auto transform = GetTransform(entity.GetTransformation()).Invert();
+  transform = matrix_ * transform;
+  transform = GetTransform(entity.GetTransformation()) * transform;
+  snapshot->transform = transform * snapshot->transform;
```
Comment (bdero) [inline impeller/entity/contents/filters/matrix_filter_contents.cc:35]: Here and below: Can we throw this arithmetic on one line to make the multiplication order easier to read?

```
snapshot->transform = entity.GetTransformation() * matrix_ * entity.GetTransformation().Invert() * snapshot->transform;
```

File: impeller/entity/contents/filters/matrix_filter_contents.cc [hunk 2]
```diff
@@ -25,7 +24,35 @@ std::optional<Snapshot> MatrixFilterContents::RenderFilter(
     const Entity& entity,
     const Matrix& effect_transform,
     const Rect& coverage) const {
-  return inputs[0]->GetSnapshot(renderer, entity);
+  auto snapshot = inputs[0]->GetSnapshot(renderer, entity);
```
Comment (bdero) [inline impeller/entity/contents/filters/matrix_filter_contents.cc:27]: The prior implementation rendered out the child filter at the final on-screen bounds resolution. This implementation no longer does so -- which I believe is correct: https://fiddle.skia.org/c/6cbb551ab36d06f163db8693972be954. I'm not sure if you intended to fix this particular issue, but thought I'd mention that you incidentally fixed it. :)

**Review by bdero — APPROVED** [review]
Thanks! Just minor comments. I think this solves all of the lingering problems -- including the fact the matrix is only supposed to apply to the geometry _after_ the filter is rendered rather than before.

File: impeller/entity/contents/filters/matrix_filter_contents.cc [hunk 3]
```diff
@@ -25,7 +24,35 @@ std::optional<Snapshot> MatrixFilterContents::RenderFilter(
...
+  if (!snapshot.has_value()) {
+    return std::nullopt;
+  }
+
+  auto transform = GetTransform(entity.GetTransformation()).Invert();
```
[the remark this answers is not in the document]
Comment (ColdPaleLight) [reply impeller/entity/contents/filters/matrix_filter_contents.cc]: Ahh, I see. Done.

File: impeller/entity/contents/filters/matrix_filter_contents.cc (hunk 1 above)
Comment (ColdPaleLight) [reply impeller/entity/contents/filters/matrix_filter_contents.cc]: Done.

**Pull request merged**